File: mooneto/cala.py
"""Cala client: turns a natural-language legal question into structured claims."""
import os
import urllib.request
import json

API = "https://api.cala.ai"


class CalaError(RuntimeError):
    pass


def _post(path: str, payload: dict, timeout: int = 90) -> dict:
    key = os.environ.get("CALA_API_KEY")
    if not key:
        raise CalaError("CALA_API_KEY is not set")
    req = urllib.request.Request(
        f"{API}{path}",
        data=json.dumps(payload).encode(),
        headers={"X-API-KEY": key, "Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.loads(r.read())


def _sources_for(context_by_id: dict, ref_ids: list) -> list:
    """Flatten the origins of the referenced context entries into unique sources."""
    out, seen = [], set()
    for rid in ref_ids:
        for origin in context_by_id.get(rid, {}).get("origins", []):
            doc = origin.get("document") or origin.get("source") or {}
            url = doc.get("url")
            if url and url not in seen:
                seen.add(url)
                out.append({"name": doc.get("name") or url, "url": url})
    return out


def ask(question: str) -> dict:
    """Query Cala and normalise the response into claims, countries and laws."""
    raw = _post("/v1/knowledge/search", {"input": question})
    context_by_id = {c["id"]: c for c in raw.get("context", [])}

    claims = [
        {
            "text": item["content"],
            "sources": _sources_for(context_by_id, item.get("references", [])),
        }
        for item in raw.get("explainability", [])
    ]

    entities = raw.get("entities", [])
    by_type = lambda t: [e["name"] for e in entities if e.get("entity_type") == t]

    return {
        "question": question,
        "answer": raw.get("content", ""),
        "claims": claims,
        "countries": by_type("Country"),
        "laws": by_type("Law"),
    }

File: lib/mooneto/cala.ts
/** Cala client: turns a natural-language legal question into structured claims. */

const API = "https://api.cala.ai"

type Origin = { source?: { name?: string; url?: string }; document?: { name?: string; url?: string } }
type ContextEntry = { id: string; content: string; origins?: Origin[] }
type Explain = { content: string; references?: string[] }
type Entity = { id: string; name: string; entity_type: string }

export type Source = { name: string; url: string }
export type Claim = { text: string; sources: Source[] }
export type CalaResult = {
  question: string
  answer: string
  claims: Claim[]
  countries: string[]
  laws: string[]
}

function sourcesFor(byId: Map<string, ContextEntry>, refs: string[] = []): Source[] {
  const out: Source[] = []
  const seen = new Set<string>()
  for (const ref of refs) {
    for (const origin of byId.get(ref)?.origins ?? []) {
      const doc = origin.document ?? origin.source ?? {}
      if (doc.url && !seen.has(doc.url)) {
        seen.add(doc.url)
        out.push({ name: doc.name ?? doc.url, url: doc.url })
      }
    }
  }
  return out
}

export async function ask(question: string): Promise<CalaResult> {
  const key = process.env.CALA_API_KEY
  if (!key) throw new Error("CALA_API_KEY is not set")

  // Cala occasionally returns a transient 5xx. A live demo cannot afford that,
  // so retry server errors with a short backoff.
  let res: Response | null = null
  for (let attempt = 0; attempt < 3; attempt++) {
    res = await fetch(`${API}/v1/knowledge/search`, {
      method: "POST",
      headers: { "X-API-KEY": key, "Content-Type": "application/json" },
      body: JSON.stringify({ input: question }),
    })
    if (res.ok) break
    if (res.status < 500) throw new Error(`Cala returned ${res.status}`)
    console.warn(`[cala] ${res.status} on attempt ${attempt + 1}, retrying`)
    await new Promise((r) => setTimeout(r, 800 * (attempt + 1)))
  }
  if (!res || !res.ok) throw new Error(`Cala unavailable (${res?.status ?? "no response"})`)

  const raw = await res.json()
  const byId = new Map<string, ContextEntry>(
    (raw.context ?? []).map((c: ContextEntry) => [c.id, c])
  )
  const entities: Entity[] = raw.entities ?? []
  const namesOfType = (t: string) =>
    entities.filter((e) => e.entity_type === t).map((e) => e.name)

  return {
    question,
    answer: raw.content ?? "",
    claims: (raw.explainability ?? []).map((e: Explain) => ({
      text: e.content,
      sources: sourcesFor(byId, e.references),
    })),
    countries: namesOfType("Country"),
    laws: namesOfType("Law"),
  }
}
